### PR TITLE
CLN: Catch more specific exception in groupby.ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
Lots of these to whittle down.

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -212,9 +212,12 @@ def apply(self, f, data, axis=0):
                 # This Exception is also raised if `f` triggers an exception
                 # but it is preferable to raise the exception in Python.
                 pass
-            except Exception:
-                # raise this error to the caller
-                pass
+            except TypeError as err:
+                if "Cannot convert" in str(err):
```
[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/core/groupby/ops.py]: I may have asked this wrong but I don't think we are on the same page. So I was thinking to keep catching `TypeError` but was questioning if we need the conditional block therein, as it diverges slightly from the pre-existing behavior.

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: > if we need the conditional block therein, as it diverges slightly from the pre-existing behavior.

Changing the behavior is intentional.  ATM we are `pass`ing on everything, and I want to `pass` on a narrow set of `TypeError`s.

File: pandas/core/groupby/ops.py [hunk 2]
```diff
@@ -212,9 +212,12 @@ def apply(self, f, data, axis=0):
...
+                    # via apply_frame_axis0 if we pass a non-ndarray
+                    pass
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: looks fine, does this change any perf? IOW this looks like this is now taking a path that previously we raised (and then likely did an .apply on)

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: Not sure.  Between this, #27909, and a not-yet-pushed branch that fixes incorrect exception handling in cython_agg_block, I'm pretty sure we'll end up falling back to python-space less often, but it isn't obvious what the individual changes affect perf-wise.

**Comment by jreback** [comment]
lgtm. can you rebase as I think we merged some other groupby/ops stuff. (just to be sure).

**Review by WillAyd — APPROVED** [review]

**Pull request merged**